File: tools/ghidra_apply_merged_symbols.java
//Apply merged symbol names to stbc.exe from tools/merged/stbc_merged_symbols.csv.
//Renames each function at its address and attaches a plate comment carrying
//owning_class / ancestors / binding_kind / join_key. Java version so it runs in
//stock Ghidra with no Jython/PyGhidra. Run from Script Manager (category STBC).
//@category STBC
//@author STBC-RE
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.CodeUnit;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.symbol.SourceType;

public class ghidra_apply_merged_symbols extends GhidraScript {

    // Fallback if the CSV can't be located next to the script and no arg given.
    private static final String CSV_DEFAULT =
        "C:\\Users\\mward\\Documents\\Projects\\STBC-Reverse-Engineering"
      + "\\STBC-Reverse-Engineering\\tools\\merged\\stbc_merged_symbols.csv";

    private File resolveCsv() {
        String[] args = getScriptArgs();
        if (args != null && args.length >= 1 && args[0] != null && !args[0].isEmpty()) {
            return new File(args[0]);
        }
        // Prefer the copy next to this script: <tools>/merged/stbc_merged_symbols.csv
        try {
            File toolsDir = getSourceFile().getParentFile().getFile(false);
            File beside = new File(toolsDir, "merged/stbc_merged_symbols.csv");
            if (beside.isFile()) {
                return beside;
            }
        } catch (Exception e) {
            // fall through to default
        }
        return new File(CSV_DEFAULT);
    }

    // Split a CSV line into exactly the header's field count. Our fields never
    // contain commas (identifiers, class names, and '|'-separated ancestors),
    // so a plain split is safe.
    private String[] fields(String line, int n) {
        String[] parts = line.split(",", -1);
        if (parts.length < n) {
            String[] padded = new String[n];
            System.arraycopy(parts, 0, padded, 0, parts.length);
            for (int i = parts.length; i < n; i++) padded[i] = "";
            return padded;
        }
        return parts;
    }

    private String buildComment(String category, String owning, String pyClass,
                                String binding, String ancestors, String joinKey) {
        StringBuilder sb = new StringBuilder();
        if (!category.isEmpty()) sb.append("category=").append(category).append("  ");
        if (!owning.isEmpty())   sb.append("class=").append(owning).append("  ");
        if (!pyClass.isEmpty() && !pyClass.equals(owning))
            sb.append("py_class=").append(pyClass).append("  ");
        if (!binding.isEmpty())  sb.append("binding=").append(binding).append("  ");
        if (!joinKey.isEmpty())  sb.append("join=").append(joinKey);
        String line1 = sb.toString().trim();
        if (!ancestors.isEmpty()) {
            return line1 + "\n  ancestors: " + ancestors.replace("|", " -> ");
        }
        return line1;
    }

    @Override
    public void run() throws Exception {
        File csv = resolveCsv();
        if (!csv.isFile()) {
            println("[ERR] merged CSV not found: " + csv.getAbsolutePath());
            return;
        }
        println("reading " + csv.getAbsolutePath());

        FunctionManager fm = currentProgram.getFunctionManager();
        Listing listing = currentProgram.getListing();

        int nTotal = 0, nRenamed = 0, nSame = 0, nCommented = 0, nCreated = 0;
        int nNoFunc = 0, nBadAddr = 0, nErr = 0;

        BufferedReader br = new BufferedReader(new FileReader(csv));
        try {
            String header = br.readLine(); // skip header
            if (header == null) {
                println("[ERR] CSV is empty");
                return;
            }
            String line;
            while ((line = br.readLine()) != null) {
                if (line.trim().isEmpty()) continue;
                nTotal++;
                if (monitor.isCancelled()) break;

                String[] f = fields(line, 9);
                String addrStr  = f[0].trim();
                String name     = f[1].trim();
                String category = f[2].trim();
                String owning   = f[3].trim();
                String pyClass  = f[4].trim();
                String binding  = f[6].trim();
                String ancestors= f[7].trim();
                String joinKey  = f[8].trim();

                if (addrStr.isEmpty() || name.isEmpty()) { nErr++; continue; }

                String hex = addrStr.toLowerCase().startsWith("0x")
                           ? addrStr.substring(2) : addrStr;
                Address addr;
                try {
                    addr = toAddr(hex);
                } catch (Exception e) {
                    addr = null;
                }
                if (addr == null) { nBadAddr++; continue; }

                Function func = fm.getFunctionAt(addr);
                boolean created = false;
                if (func == null) {
                    // Not yet a defined function -- typical for SWIG wrappers /
                    // NiRTTI factories only reached via data pointer tables.
                    // Disassemble the bytes and carve out a function here.
                    try {
                        disassemble(addr);
                        func = createFunction(addr, name);
                    } catch (Exception e) {
                        func = null;
                    }
                    if (func == null) { nNoFunc++; continue; }
                    created = true;
                    nCreated++;
                }

                try {
                    if (created) {
                        // name already set by createFunction; nothing to rename
                    } else if (func.getName().equals(name)) {
                        nSame++;
                    } else {
                        func.setName(name, SourceType.USER_DEFINED);
                        nRenamed++;
                    }
                    String comment = buildComment(category, owning, pyClass,
                                                  binding, ancestors, joinKey);
                    if (!comment.isEmpty()) {
                        listing.setComment(addr, CodeUnit.PLATE_COMMENT, comment);
                        nCommented++;
                    }
                } catch (Exception e) {
                    nErr++;
                    if (nErr <= 20) println("[warn] " + name + " @ " + addrStr + ": " + e);
                }
            }
        } finally {
            br.close();
        }

        println("========================================================");
        println("merged-symbol apply complete");
        println("  rows read           : " + nTotal);
        println("  functions renamed   : " + nRenamed);
        println("  functions created   : " + nCreated);
        println("  already correct     : " + nSame);
        println("  plate comments set  : " + nCommented);
        println("  addr not carvable   : " + nNoFunc);
        println("  bad address         : " + nBadAddr);
        println("  errors              : " + nErr);
        println("========================================================");
    }
}

File: tools/ghidra_apply_merged_symbols.py
# Ghidra Jython Script: Apply merged symbol names to stbc.exe
# @category STBC
# @description Renames functions at their addresses from the blended symbol
#   table (tools/merged/stbc_merged_symbols.csv) and attaches a plate comment
#   carrying owning_class / ancestors / binding_kind / join_key. This is the
#   "fresh decompilation" seed: run it, then export decompiled C.
#
#   Works in BOTH the GUI Script Manager (with stbc.exe loaded) and headless:
#     analyzeHeadless <proj_dir> <proj> -process stbc.exe \
#         -postScript ghidra_apply_merged_symbols.py <path-to-csv>
#
#   If no CSV path arg is given, falls back to MERGED_CSV_DEFAULT below.
#
# Provenance of the input: tools/merge_symbols.py blends this repo's
# address-anchored function_names_export.txt with the vendored SWIG ABI export
# under tools/abi_export/. See tools/SYMBOLS.md.

import csv
import os

from ghidra.program.model.symbol import SourceType
from ghidra.program.model.listing import CodeUnit

# Edit this if you run from the GUI and don't pass the CSV as a script argument.
MERGED_CSV_DEFAULT = \
    r"C:\Users\mward\Documents\Projects\STBC-Reverse-Engineering" \
    r"\STBC-Reverse-Engineering\tools\merged\stbc_merged_symbols.csv"


def resolve_csv_path():
    args = getScriptArgs()  # noqa: F821 (Ghidra-injected)
    if args and len(args) >= 1 and args[0]:
        return args[0]
    return MERGED_CSV_DEFAULT


def build_comment(row):
    """Compact plate comment from the ABI-enrichment columns (skip empties)."""
    bits = []
    if row.get("category"):
        bits.append("category=%s" % row["category"])
    if row.get("owning_class"):
        bits.append("class=%s" % row["owning_class"])
    if row.get("python_class") and row["python_class"] != row.get("owning_class"):
        bits.append("py_class=%s" % row["python_class"])
    if row.get("binding_kind"):
        bits.append("binding=%s" % row["binding_kind"])
    if row.get("join_key"):
        bits.append("join=%s" % row["join_key"])
    line1 = "  ".join(bits)
    anc = row.get("ancestors") or ""
    if anc:
        return line1 + "\n  ancestors: " + anc.replace("|", " -> ")
    return line1


def main():
    csv_path = resolve_csv_path()
    if not os.path.isfile(csv_path):
        print("[ERR] merged CSV not found: %s" % csv_path)
        return

    fm = currentProgram.getFunctionManager()          # noqa: F821
    af = currentProgram.getAddressFactory()           # noqa: F821
    listing = currentProgram.getListing()             # noqa: F821

    n_total = 0
    n_renamed = 0
    n_same = 0
    n_no_func = 0
    n_bad_addr = 0
    n_err = 0
    n_commented = 0

    fh = open(csv_path, "rb")   # Jython csv wants bytes-mode file handle
    try:
        reader = csv.DictReader(fh)
        for row in reader:
            n_total += 1
            addr_str = (row.get("address") or "").strip()
            new_name = (row.get("name") or "").strip()
            if not addr_str or not new_name:
                n_err += 1
                continue

            # Normalize "0x00609700" -> "00609700" for the address factory.
            hexpart = addr_str[2:] if addr_str.lower().startswith("0x") else addr_str
            try:
                addr = af.getAddress(hexpart)
            except Exception:
                addr = None
            if addr is None:
                n_bad_addr += 1
                continue

            func = fm.getFunctionAt(addr)
            if func is None:
                # Address isn't the entry of a defined function. Rather than
                # invent one, record it; a follow-up pass can create functions.
                n_no_func += 1
                continue

            try:
                cur = func.getName()
                if cur == new_name:
                    n_same += 1
                else:
                    func.setName(new_name, SourceType.USER_DEFINED)
                    n_renamed += 1

                comment = build_comment(row)
                if comment:
                    listing.setComment(addr, CodeUnit.PLATE_COMMENT, comment)
                    n_commented += 1
            except Exception, e:   # Jython 2 syntax
                n_err += 1
                if n_err <= 20:
                    print("[warn] %s @ %s: %s" % (new_name, addr_str, e))
    finally:
        fh.close()

    print("=" * 56)
    print("merged-symbol apply complete")
    print("  rows read           : %d" % n_total)
    print("  functions renamed   : %d" % n_renamed)
    print("  already correct     : %d" % n_same)
    print("  plate comments set  : %d" % n_commented)
    print("  addr not a function : %d" % n_no_func)
    print("  bad address         : %d" % n_bad_addr)
    print("  errors              : %d" % n_err)
    print("=" * 56)


main()
